src/main/scala/mesosphere/marathon/core/launchqueue/impl/ReviveOffersStreamLogic.scala

```diff
@@ -0,0 +1,132 @@
+      case (current, Left(InstanceUpdated(updated, _, _))) => current.withInstanceUpdated(updated)
+      case (current, Left(InstanceDeleted(deleted, _, _))) => current.withInstanceDeleted(deleted)
+      case (current, Right(Delayed(configRef))) => current.withDelay(configRef)
+      case (current, Right(NotDelayed(configRef))) => current.withoutDelay(configRef)
+    }
+  }
+
+  /**
+    * A resident instances was deleted and requires a revive.
+    *
+    * Currently Marathon uses [[InstanceTrackerDelegate.forceExpunge]] when a run spec with resident instances
+    * is removed. Thus Marathon looses all knowledge of any reservations to these instances. The [[OfferMatcherReconciler]]
+    * is supposed to filter offers for these reservations and destroy them if no related instance is known.
+    *
+    * A revive call to trigger an offer with said reservations to be destroyed should be emitted. There is no
+    * guarantee that the reservation is destroyed.
+    *
+    * @param current The current accumulated state.
+    * @param previous The accumulated state of the last check.
+    * @return true if there is a new force expunged resident instance, false otherwise.
+    */
+  def shouldReconcileReservation(current: ReviveOffersState, previous: ReviveOffersState): Boolean =
+    current.forceExpungedInstances > previous.forceExpungedInstances
+
+  /**
+    * This flow emits a [[Suppress]] or a [[Revive]] based on the last two states emitted by [[reviveStateFromInstancesAndDelays]]
+    *
+    * @return The actual flow.
+    */
+  val suppressOrReviveFromDiff: Flow[ReviveOffersState, Op, NotUsed] = Flow[ReviveOffersState]
+    .sliding(2)
+    .mapConcat {
+      case Seq(previous, current) =>
+
+        val diffScheduled = current.scheduledInstancesWithoutBackoff -- previous.scheduledInstancesWithoutBackoff
+        val diffTerminal = current.terminalReservations -- previous.terminalReservations
+
+        if (diffScheduled.nonEmpty || diffTerminal.nonEmpty) {
+          logger.info(s"Revive because new scheduled $diffScheduled or terminal $diffTerminal.")
+          // There's a very small chance that we decline an offer in response to a revive for an instance not yet registered
+          // with the TaskLauncherActor. To deal with the rare case this happens, we just repeat the last suppress / revive
+          // after a while.
+          List(Revive, Revive, Revive)
+        } else if (shouldReconcileReservation(current, previous)) {
+          logger.info(s"Revive to trigger reservation reconciliation.")
+          List(Revive)
+        } else if (current.isEmpty) {
+          logger.info(s"Suppress because there are no pending instances right now.")
+          List(Suppress)
+        } else {
+          logger.info("Nothing changed in last frame.")
+          Nil
+        }
+      case _ =>
+        logger.warn("Did not receive two elements; end of stream detected")
```

Comment:
Please provide context to this and others logging lines e.g.:
```suggestion
        logger.warn("Suppress/Revive: did not receive two elements; end of stream detected")
```